src/output/plugins/OSXOutputPlugin.cxx

```diff
@@ -220,43 +221,116 @@ osx_output_set_device(OSXOutput *oo, Error &error)
 	return ret;
 }
 
+
+/*
+	This function (the 'render callback' osx_render) is called by the
+	OS X audio subsystem (CoreAudio) to request audio data that will be
+	played by the audio hardware. This function has hard time constraints
+	so it cannot do IO (debug statements) or memory allocations.
+	
+	The caller (i.e. CoreAudio) requests a specific number of
+	audio frames (in_number_frames) to be rendered into a
+	collection of output buffers (buffer_list). Depending on the
+	number of output buffers the render callback has to interleave
+	or de-interleave audio data to match the layout of the output
+	buffers. The intput buffer is always interleaved. In practice,
+	it seems this callback always gets a single output buffer
+	meaning that no de-interleaving actually takes place. For the
+	sake of correctness this callback allows for de-interleaving
+	anyway, and calculates the expected output layout by examining
+	the output buffers.
+	
+	The input buffer is a DynamicFifoBuffer. When a
+	DynamicFifoBuffer contains less data than we want to read from
+	it there is no point in doing a second Read(). So in the case
+	of insufficient audio data in the input buffer, this render
+	callback will emit silence into the output buffers (memset
+	zero).
+*/
+
 static OSStatus
 osx_render(void *vdata,
 	   gcc_unused AudioUnitRenderActionFlags *io_action_flags,
 	   gcc_unused const AudioTimeStamp *in_timestamp,
 	   gcc_unused UInt32 in_bus_number,
-	   gcc_unused UInt32 in_number_frames,
+	   UInt32 in_number_frames,
 	   AudioBufferList *buffer_list)
 {
+	AudioBuffer *output_buffer = nullptr;
+	size_t output_buffer_frame_size, dest;
+
 	OSXOutput *od = (OSXOutput *) vdata;
-	AudioBuffer *buffer = &buffer_list->mBuffers[0];
-	size_t buffer_size = buffer->mDataByteSize;
+	DynamicFifoBuffer<uint8_t> *input_buffer = od->buffer;
+	assert(input_buffer != nullptr);
+
+	/*
+		By convention when interfacing with audio hardware in CoreAudio,
+		in_bus_number equals 0 for output and 1 for input. Because this is an
+		audio output plugin our in_bus_number should always be 0.
+	*/
+	assert(in_bus_number == 0);
+
+	unsigned int input_channel_count = od->asbd.mChannelsPerFrame;
+	unsigned int output_channel_count = 0;
+	for (unsigned int i = 0 ; i < buffer_list->mNumberBuffers; ++i) {
+		output_buffer = &buffer_list->mBuffers[i];
+		assert(output_buffer->mData != nullptr);
+		output_channel_count += output_buffer->mNumberChannels;
+	}
+	assert(output_channel_count == input_channel_count);
 
-	assert(od->buffer != nullptr);
+	size_t input_buffer_frame_size = od->asbd.mBytesPerFrame;
+	size_t sample_size = input_buffer_frame_size / input_channel_count;
 
+	// Acquire mutex when accessing input_buffer
 	od->mutex.lock();
 
-	auto src = od->buffer->Read();
-	if (!src.IsEmpty()) {
-		if (src.size > buffer_size)
-			src.size = buffer_size;
-
-		memcpy(buffer->mData, src.data, src.size);
```

Comment:
This old memcpy assumed that the output buffer was interleaved in the exact same way as the input buffer.